pwndbg/storage.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Provides persistent volatile and non-volatile storage of the results of
+a function call.  Volatile storage uses the temporary directory, and is
+generally cleared each reboot, or as the OS dictates.
+
+Cached results are stored per-architecture, i.e. arm and i386 caches
+are independent.  Cached results are also stored per-argument.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import functools
+import inspect
+import os
+import pickle
+import shelve
+import tempfile
+import time
+
+import pwndbg.arch
+import pwndbg.config
+
+nonvolatile = pwndbg.config.Parameter('pwndbg-cache-nonvolatile', os.path.expanduser('~/.pwndbg'), 'Non-volatile cache directory for Pwndbg storage')
+volatile = pwndbg.config.Parameter('pwndbg-cache-volatile', tempfile.gettempdir(), 'Volatile cache directory for Pwndbg storage')
+
+# Ensure there is only one Shelf instance per database, so that
+# we don't get weird things with multiple instances / concurrent writes / etc.
+shelves = {
+    nonvolatile: None,
+    volatile: None
+}
+
+@pwndbg.config.Trigger([volatile, nonvolatile])
+def update():
+    global shelves
+
+    for db in (volatile, nonvolatile):
+        database_path = os.path.join(str(db), 'storage')
+
+        # Ensure the path exists, then open the database
+        if not os.path.isdir(database_path):
```

Comment:
The `database_path` must be a path to a file (might be non existing) otherwise `shelve.open` fails:
```
$ gdb
Defined pretty print function (pp) and libc (CDLL)
Loaded 108 commands. Type pwndbg [filter] for a list.
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/gdbinit.py", line 36, in <module>
    import pwndbg # isort:skip
  File "/home/dc/installed/pwndbg/pwndbg/__init__.py", line 69, in <module>
    import pwndbg.storage
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 53, in <module>
    update()
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 50, in update
    protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.6/shelve.py", line 243, in open
    return DbfilenameShelf(filename, flag, protocol, writeback)
  File "/usr/lib/python3.6/shelve.py", line 227, in __init__
    Shelf.__init__(self, dbm.open(filename, flag), protocol, writeback)
  File "/usr/lib/python3.6/dbm/__init__.py", line 94, in open
    return mod.open(file, flag, mode)
_gdbm.error: [Errno 21] Is a directory
```

I'd go with:
```python
for db in map(str, (volatile, nonvolatile)):
    ...
```